### DISCUSSION TITLE
Property view

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#898] [Q&A]
Hello, 
I added a tool palette to my GLSP view and fill it with different elements, I can add those elements to my view.
My project is based on the `node-json-theia` example.

I want to open a `panel`  when double-clicking an element.
this panel contains the element properties or some of them which I can edit .

Am wondering if there are some docs concerning this task or if you could explain to me the way to develop it. 

Thanks.

**Comment by tortmayr** [comment]
Hi @Adem-Kri,
there is no documentation available that explicitly covers this concrete task.
Properties views/pannels are typically integrated in one of two ways
 1. directly as part of the diagram using the [UIExtension](https://www.eclipse.org/glsp/documentation/ui-extensions/) API
 2. or as part of the surrounding application frame (i.e. a Theia widget in your case)
 
The first approach has the advantage that it's generic code that can be easily reused if you want integrate your GLSP editor into another application at some point (e.g. VSCode or standalone). The second approach on the other hand guarantees a seamless integration with the existing UI and will probably result in a overall better UX. However, this also means that the code probably can not be (easily) reused in other application integrations.

As far as I know @rsoika has successfully implemented a Property panels using Approach 1 (UI extensions). Maybe he is willing to share his insights 😉.
@ndoschek Could you maybe give some details on how to implement native property views for GLSP in Theia?

**Comment by rsoika** [comment]
Hi @Adem-Kri, yes I have implemented a solution following approach 1 to be able to use the same functionality in Theia and in VS-Code.
Even as I added JsonForms in addition, the general setup and code needed for this additional feature is quite simple. 
You can see the full source code [here](https://github.com/imixs/open-bpmn/blob/master/open-bpmn.glsp-client/open-bpmn-properties/src/bpmn-property-panel.tsx). 

![image](https://user-images.githubusercontent.com/1339124/216937935-d5047a0a-bfab-4d8f-a0b8-4faf0713fc7b.png)


The UI Extension is than bound to the `ContainerModule`. In this way the panel is a separate independent module or build block.

```javascript
export default function createBPMNDiagramContainer(widgetId: string): Container {
    const container = createDiagramContainer(bpmnDiagramModule, bpmnPropertyModule);
    overrideViewerOptions(container, {
        baseDiv: widgetId,
        hiddenDiv: widgetId + '_hidden'
    });
    return container;
}
```
As you can see in the code, I implemented also a `SelectionListener` to react on element-clicks and update the content of my panel.

The double-click feature is also something I wanted in my implementation, but for some reason I failed to implement an additional mouse listener. Maybe if you succeeded, you can share your solution too.

You can see a live demo here: https://open-bpmn.imixs.org

**Reply by Adem-Kri** [reply]
Hello @rsoika, 
Thank you for your answer; I found it really useful.

But I have an issue with `createDiagramContainer` it keeps telling me
```
src/di.config.ts:28:5 - error TS2305: Module '"@eclipse-glsp/client"' has no exported member 'createDiagramContainer'.

28     createDiagramContainer
       ~~~~~~~~~~~~~~~~~~~~~~

Found 1 error.
```

Even though I tried `yarn add @eclipse-glsp/client`, 

do you have a hint about this, 

thank you in advance.

**Reply by tortmayr** [reply]
@Adem-Kri 
What GLSP version are you using?
The `createDiagramContainer` method just got introduced recently and is only part of the nightly/Release Candiate versions. 
In the 1.0.0 Release and below the method is called `createClientContainer`.

**Reply by Adem-Kri** [reply]
@tortmayr 
I am using `"@eclipse-glsp/client": "^1.0.0",`

how could I upgrade it, and would it make confusion inside the project ?

In addition, in case I used `createDiagramContainer`, should I delete the function using `createClientContainer`

here's my code:

```
export default function createContainer(widgetId: string): Container {
    const container = createClientContainer(DiagramModule);

    overrideViewerOptions(container, {
        baseDiv: widgetId,
        hiddenDiv: widgetId + '_hidden',
        needsClientLayout: true
    });

    return container;
}


export function createDiagramContainer(widgetId: string): Container{
    const container = createDiagramContainer(DiagramModule, PropertyModule);
    overrideViewerOptions(container, {
        baseDiv: widgetId,
        hiddenDiv: widgetId + '_hidden'
    });
    return container
}
```

**Reply by tortmayr** [reply]
I would not recommend to upgrade unless there is some functionality that you absolutely need. The nightly versions contains some API breaks so updating is no as straight forward.

For 1.0.0 just keep using `createClientContainer`.

**Reply by Adem-Kri** [reply]
I'm not trying to be rude by asking too many questions.

But is there an alternative to using `createDiagramContainer`, as proposed by @rsoika ?

I tried upgrading GLSP to 1.1.0, although @tortmayr recommended not to, and got a bunch of new errors.

is there a way to implement it using 1st approach and bypassing the `createDiagramContainer` issue?

thank you.

**Reply by tortmayr** [reply]
Sorry, maybe I wasn't clean clear enough.
You can reuse the solution that @rsoika proposed and simply replace `createDiagramContainer` with `createClientContainer`. Implementation-wise both functions do the same thing.

So if you adapt your second snippet like this:

```ts
export function createDiagramContainer(widgetId: string): Container{
    const container = createClientContainer(DiagramModule, PropertyModule);
    overrideViewerOptions(container, {
        baseDiv: widgetId,
        hiddenDiv: widgetId + '_hidden'
    });
    return container
}
```

it should work for  GLSP 1.0.0.

**Comment by ndoschek** [comment]
Hi @Adem-Kri,

if you would like to use Theia's generic property view (i.e. that is a separate widget that can also show file information and is repositionable and closable) please checkout the following basic documentation in Theia's concept documentation: https://theia-ide.org/docs/property_view

Also, in recent discussions (https://github.com/eclipse-glsp/glsp/discussions/860 and https://github.com/eclipse-glsp/glsp/discussions/888),  that use `java-emf-theia` as project template however, we came up with an example branch to integrate the Theia property view with GLSP: https://github.com/eclipse-glsp/glsp-examples/commit/70fdb1ab998c8e1e3051bf1b3266d1a21e44b2c3
Please note that is a POC, and does not contain the actual editing of values yet.

HTH and best regards from Vienna,
Nina
